### ISSUE TITLE
upgrade to buster / upstream things like xkb config

### ISSUE THREAD

**Issue by drhpc** [gemian/gemian#26]
I am looking into making some current use of the Gemin PDA at all. The goal: An SSH terminal that utilizes SSH keys on USB security keys managed by libfido2 (ed25519-sk), uplink via wifi. I got a 4G model, but I'll use it without SIM. Support for this rather fringe on Android (might be there eventually), so I figured a 'normal' Linux on the neat little device would do the trick. Would fit my use case of a dedicated SSH terminal for admin purposes.

I managed to flash the debian stretch firmware again, also got some upgrades from archive.debian.org. Just some work to get libreoffice out of the way.

For the heck of it, I also managed to do an apt upgrade to buster. But the gemian.thinkglobally.org repos are down, and matching releases of the adapted packages would be missing, anyway. I would be fine with a rather basic experience, no productivity apps. An obvious missing piece is the xkeyboard-config data for the custom keyboard. Any reason why that is not upstreamed? Seems to be a harmless addition. How much special sauce is there to get to the desktop UI at all? (And: Are there kernel VTs at all? The documentation I can find is rather sparse, as someone not involved in Android-type device OS development otherwise.)

Of course my attempt at upgrading results in a boot stuck at the splash image for Gemian. Something behind the scenes does not continue.  I am stuck at the boot screen and I am missing any special knowledge how to debug this. Is there a serial console? Can I press a magic key to see the kernel console buffer?

I see some things mentioned about EGL libs/config in the wiki. I forgot to check those files before reboot … would that have helped? I can go through the upgrade process again. But has anyone ever booted a Debian buster with the aging 3.18 kernel on the Gemini? Should I just stay with stretch and build more modern userspace software on top? I could get away with compiling libfido2 and openssh for my purposes. But one showstopper is the aged wifi stack: I am unable to connect to an eduroam network. That is rather limiting. I don't mind outdated desktop software otherwise. But of course, it would be cool to get the device supported properly and updated via standard debian repos for a few years.

So, is there any life left in this project for the Gemini for updated Debian? Have there been attempts at upstreaming the package hanges?

Edit: I see https://github.com/gemian/gemian/wiki/BuildServer#build-order now with what seems like a list of packages necessary to be custom-built for Gemian (first among those a glibc with the ancient kernel headers). So is just someone needed to pull the trigger on the Jenkins server and host the resulting repo?

**Comment by adamboardman** [comment]
Feel free to try to upstream things, if you do the keyboard config then please include the Astro stuff (bookworm branch).

The existing repositories should have all the packages, though you should know that for buster we ended up with two build servers as Planet provided an additional server and packages got split between them:
```
deb https://gemian.thinkglobally.org/buster/ buster main
deb https://gemian-planet.thinkglobally.org/buster/ buster main
```

I'm currently working on bookworm for Astro, once I've got that working then the idea is to roll that back to the Cosmo and then after that to the Gemini (probably - and more likely if folk are actually interested). I was on strike for 15 months due to Planet refusing to answer questions - their upstream went bust and they had to re-negotiate everything it seems. For bookworm I went back to just my build server as getting planet to re-enter the repo passphrase after each reboot takes too long that its just not worth the wait.

You can get a general feed of activity:
https://github.com/orgs/gemian/dashboard

Edit: You can use a usb cable to login either with tethering (full linux access when GUI fails), stuck bootloader, or usb-uart for kernel logs:
https://github.com/gemian/gemian/wiki/Debugging

**Comment by semiprime** [comment]
> Gemini (probably - and more likely if folk are actually interested)

I would be very interested in an update of Gemian for the Gemini.

(And thanks for the work you've already done - my Gemini is Gemian-only and I use it every day.)

**Comment by drhpc** [comment]
Great to see that there is life left. Can we hope for a newer kernel build some day (Mediatek opening up?!??)?

I'm trying to access the revived repos now on the device. One stumbling block is the HTTPS setup of deb https://gemian.thinkglobally.org/stretch/ (and probably buster, too). I get 403 on a web browser (intentionally?) and the Gemini gets either a certificate from the future (if I didn't set the time yet) or an expired one (if I did set the time).

I'll see how far I get and will probably open another issues for specific things.

PS: How do I sync the time to the hardware clock? hwclock doesn't seem  to work. Is NTP on boot the only way?

**Comment by adamboardman** [comment]
I think you can just avoid using https:
```
deb [trusted=yes] http://gemian.thinkglobally.org/stretch/ stretch main
```

**Comment by drhpc** [comment]
Somehow I cannot. The server hard-redirects to https via 301. I am struggling to disable the cert validation. Hm. Also, is it expected that gemian.thinkglobally.org/stretch and buster give me 403 (toplevel) or 404 (anything below) in the browser? The gemian-planet.thinkglobally.org/buster shows me someting, but only after accepting the wrong certificate (which is for pulsar.planetcom.co.uk, not thinkglobally.org).

I'll try various things, as I'd like to upgrade to latest state of stretch before upgrading to buster — maybe I have to drop that.

But is the stretch repo definitely up? Same for buster? The server seems to be configured rather restrictively, so that I am not sure if necessary bits are really there. So far I only have seen the new split repo on planetcom's server.

Edit: apt complains that there is no Release file for stretch in any case. Is there supposed to be one? Options to overcome the security check so far fail (even trusted=yes,allow-downgrade-to-insecure=yes).

Edit 2:  Somehow I got the option syntax for individual repos wrong, I guess. What does work is `apt update -o Acquire::AllowInsecureRepositories=true`.

**Comment by adamboardman** [comment]
The following file exists: https://gemian.thinkglobally.org/stretch/dists/stretch/InRelease

The planet server is really pulsar though their cert should also include the gemian-planet name too I would have thought, though I have no control of that side of things, best to avoid https if it is messed up. I guess my syntax is for bookworm versions as that's all I've worked with for the last while.

**Comment by drhpc** [comment]
Thanks. I managed to updated to latest stretch stuff, with AllowInsecureRepositories. Now trying to get to buster. It's happily downloading packages. I still have to add those https exceptions. Would be nice if the certs could be fixed.

I got USB networking for a debugging view should X11 not work. I'm staying tuned for further updates … but I wonder: Are we looking at an image install of bookworm, as in the upgrade path we're missing bullseye for Gemini? Also, I suspect we'll need a new kernel at some point?


PS: With NetworkManager, connecting to an WPA enterprise network (eduroam) does work also with stretch. So I guess I could get some way with that, but of course a supported distro feels better …
